### ISSUE TITLE
Proposal to replace Macro Annotations

### ISSUE THREAD

**Issue by DavidDudson** [scalameta/scalameta#1182]
Hi all,

I need to use macros similar to how scalameta does under the hood, in order to generate AST information. Given a specific case class, I also want to generate instances of Eq, Show, Functor etc. into companions.

I need this sooner rather then later, so I would like to help out by implementing this myself if the rest of the scalameta team is happy with this (I'm going to write this anyway for personal use, so I might aswell expose it to the rest of the community)

Below is a table of what is _possible_ with code gen, not necessarily what will be implemented.

Feature | Code gen | Macro annotations
-----------| ------------ | -------------
Syntactic | Yes | Yes
Enclosing scope inspection | Maybe** | Yes
Same return | Yes | Yes
Different return | Yes | No
Composition | Yes | No***
Extensions | Yes | Yes
Manipulation | Yes | Yes
Nested | Yes | Yes
Duplicate | Yes | Yes
Multiples | Yes | Yes
Recursive | Yes | ???
Source Maps | Yes | No
Diffs | Yes | No
See outputs | Yes | No
Compiler plugin | No | Yes
IDE friendly | Yes | No
Value Parameters | Yes | Yes
Type Parameters | Yes | Yes
Backend agnostic | Yes | No

Note: The above is largely incomplete

- * It would be possible to provide a semantic database of sources not modifiable by code gen (eg. out of the current SBT project, so it's plausible to be able to add library semantics, and possibly other projects)
- ** Syntactic only, this would involve populating the parent field of the tree, initially, the answer to this would be no, but this is not a hard constraint.
- *** It it possible, by converting statics into methods.
??? = cannot remember whether it is valid

### Syntactic

By leveraging scalameta, we can ensure all tree's are syntactically valid. This does mean the code will compile. There is no semantic support such as fully qualified names, nor can we resolve any types.

Example: 

If we wanted to generate a very trivial equals method.

```scala
 @GenEquals
 class Point(val x: Int, val y: Int)
```

could become

```scala 
class Point(val x: Int, val y: Int) {
  def equals(a: Any): Boolean = {
    a match {
       case Point(ox, oy) => 
           ox == x && oy == y
       case _ => false
    }
  }
}
```

### Enclosing

Due to being a post-order traversal, we know anything that has already been traversed has already been fully expanded. We know that extension generators do not tamper with existing information, thus are safe to introspect. Thus the only issues we have, are with not-yet-traversed tree's that have Manipulation based generators. I propose we only allow introspection of the _source code_ of enclosing scopes, any generated code is off limits, this dramatically simplifies how it can implemented and will improve performance.

### Same return

These generators take an input and output of the same type.

eg. `Defn.Def => Defn.Def`

### Different return

These generators can optionally alter the return type.

eg `Defn.Def => Defn.Val`

### Composition

The ability to compose multiple generators into a single one. 

This will initially only be available to extension generators.

### Extensions

Extend existing code.

Have the type `A => Seq[Stat]`

They can _only_ generate new methods etc. for existing trees.

In theory, these would cover most use cases. 

Note: If the entire file only uses extension generator's, the entire process can potentially happen in parallel.

### Manipulators

Modify existing code, these are the "outlaws" of code generation, and what the macro annotations are based off

Have the type `A => A`, or `A => B`

### Nested

```scala
@GeneratorA
 object Foo {
    @GeneratorB
     def bar = ???
  }
```

### Multiple

```scala
@GeneratorA @GeneratorB
 object Foo
```

### Recursive

Basically, allowing a generator, to generate another generator annotation.

### Source Maps

Since we have the original and expanded tree's with correct line numbers. We can generate source maps to put errors in the right place. Any position inside synthetic code becomes an error at the position of the annotation. Given that the output... Technical jump to definition to could even end up at the resulting file, instead of the source.

### Diffs

With code gen, we can actually show diffs, like scalafmt. We could also have dry-runs etc.

### Seeing outputs

With code gen, the output is a literal file, not part of the compilation pipeline. Thus it is easy for the user just to open the resulting file, and look for problems.

### Value parameters

```scala
@SomeGenerator(1)
class Point(x: Int, y: Int)
```

### Type Parameters

```scala
@Deriving[Eq, Show]
class Point(x: Int, y: Int)
```
## Plan

1. Add a new module to scalameta, called `gen` which is tool agnostic.
2. Add a cli interface
3. Write plugins for popular build systems (Primarily SBT)

I believe it would make sense for these projects to live and evolve with scalameta.
If necessary however, they can be put in a separate repo. I do not mind either way.

### Interface

Almost identical to scalameta paradise macro annotations, the benefit of this is it makes migration from scalameta paradise almost trivial.

1. Define an annotation 
2. Annotate the definition for which code generation must happen
3. Write a method in for the generator, using the following api: https://github.com/scalacenter/macros/issues/6
4. Register generators

### Specifics on how generation happens.

Macro generation would be a post-order traversal of the original tree.
We scan the resulting tree for generators and recursively generate code until no more generators are available. Then we continue the traversal. So effectively we recursively run post-order traversals, until no more generation is necessary.

## SBT Plugin

- Compiles the generator sources
- Builds up a set of available generators
- Traverses input files and runs generators.
- Outputs to generated sources
- Excludes the appropriate sources from compilation.

I am happy to go into more detail on how this would work, but I think code examples would be explain my logic better.

PS: If work has already been done in order to do something similar let me know.

**Comment by DavidDudson** [comment]
@xeno-by @olafurpg @ShaneDelmore thoughts?

**Comment by olafurpg** [comment]
I'm glad to hear you want to work on code gen, this is an area that has been mostly unexplored with scalameta so far but I think has big potential :)

Regarding implementation, I think the best convention for end-users would be to to have separate project `app-gen` (or some other naming convention) where annotated classes are stored.

```
.
├── app
│   └── target
│       └── scala-2.12
│           └── src_managed
│               └── app
│                   └── Data.scala
└── app-gen
    └── src
        └── main
            └── scala
                └── app
                    └── Data.scala
```

`app-gen` is just like a regular project,
- it should compile 
- it can have dependencies
- it has IDE support
The only limitation  in `app-gen` is that it cannot depend on the expanded code, since the expansion hasn't happened yet.

The build plugin uses the semanticdbs from `app-gen` as input to feed the generators and emit generated code in `app/target/src_managed`.

Generators are then written in separate project, either provided as a libraryDependency or another project in the same build.
It would be cool if

> Macro generation would be a post-order traversal of the original tree.

Sounds good 👍 

> Multiple/Nested

I think we can support multiple annotations in the same program as long as we can find them statically before expansion. 

> Recursive

I think we should not permit this, since it complicates the expansion mechanism and I don't see this making things simpler. If your generator depends on the expansion of another generator, you can call that generator's function (`OtherGenerator.expand(q"val toGenerate")`) instead of expanding an annotation (`q"@OtherGenerator val toGenerate"`).

I think would be a fair limitation if annotations in the expanded code do not get expanded.

> Thus the only issues we have, are with not-yet-traversed tree's that have Manipulation based generators. I propose we only allow introspection of the source code of enclosing scopes, any generated code is off limits, this dramatically simplifies how it can implemented and will improve performance.

I'm not sure I follow, can you elaborate?

> Add a new module to scalameta, called gen which is tool agnostic.

I think this merits a separate repo, scalameta/scalagen. A smaller repo makes testing faster and more approachable for external contributors. I've heard from many that the scalameta build is quite heavy to work in.

> Source Maps

This is indeed a big challenge, one of the biggest issues with codegen is that "go to definition" jumps to the generated sources but you want to jump to the original sources.
This is possible but probably quite hard to pull off. Actually, if you're using semanticdb, in language-server I think it will automatically detect two alternatives in "go to definition" and let the user decide which one to jump to. Need to double check.

> Write a method in for the generator, using the following api: scalacenter/macros#6

Can you write a draft of this API here?

> PS: If work has already been done in order to do something similar let me know.

You might be interested in my rework of the pretty printer (TreeSyntax, `.syntax`). I was working on it a couple months ago and have been meaning to get back to it once language-server slows down. It already improves on TreeSyntax with the following

- lots of fixed roundtrip bugs where TreeSyntax produces output that would parse into a different AST. 
- supports attaching comments/trivia to tree nodes  https://github.com/olafurpg/scala-syntax/blob/19d283682167ad732dfa97cc256b870306113cdc/tests/unit/src/test/scala/org/scalafmt/tests/DocstringSuite.scala#L7
- much better performance, ~5x faster than scalafmt
- property tests against a large corpus: https://github.com/olafurpg/scala-syntax/blob/19d283682167ad732dfa97cc256b870306113cdc/tests/slow/src/test/scala/org/scalafmt/internal/IdempotencyPropertyTest.scala 
- line wrapping with configurable column width, using typelevel/paiges

**Comment by xeno-by** [comment]
This sounds like a great avenue to explore. I like scalameta/scalagen as the location for this exploration.

**Comment by DavidDudson** [comment]
> Regarding implementation, I think the best convention for end-users would be to to have separate project app-gen (or some other naming convention) where annotated classes are stored.

I was trying to decide whether it should be a separate project, or separate sourceset. Given library dependencies will probably be different. That's an issue for the SBT plugin though

> app-gen is just like a regular project,

That was my intention, glad we are on the same page! Generator libraries could also be published easily.

>> Recursive
> I think we should not permit this, since it complicates the expansion mechanism
@xeno-by is this not what the scalameta source does?

My use case for this is the following...

Have a generator for case class like behavior. lets call it `@Data`

Generating some sort of data class from another macro. lets call it `@Specialized`

```
@Specialized[Byte, Short, Int, Double]
class Point[A](x: A, y: A)
```

becomes

```
@data class Point(x: Byte, y: Byte)
@data class Point(x: Short, y: Short)
@data class Point(x: Int, y: Int)
@data class Point(x: Double, y: Double)
```

Which would then expand into a similar state that case classes do.

>>Thus the only issues we have, are with not-yet-traversed tree's that have Manipulation based generators. I propose we only allow introspection of the source code of enclosing scopes, any generated code is off limits, this dramatically simplifies how it can implemented and will improve performance.
>I'm not sure I follow, can you elaborate?
There's 2 "modes" we could use for introspection.

1. Source only.

This is the simplest approach, we provide the parsed source of the tree as an input to the generators.

2. Best effort

We provide the tree that's been expanded so far, allowing for users to order definitions in a file, in a way that they can rely on other generators expanding first. We may want to consider this later, but not initially.

Extension vs Manipulation:

When we are introspecting our enclosing scope, we have to understand what is subject to change.

If what we are introspecting has a parent that is annotated with a manipulator based annotation, we cannot guarentee anything, the manipulator can modify everything from the signature of the annotee, to it's contents.

If what we are introspecting has a parent that is annotated with a manipulator based annotation, we can guarantee that all existing code will also be in the result. We can only say that extra stats _may_ end up in the body of the annotee. This is similar to introspecting a case class, as it's equals and hash code methods etc. do not exist in the structural model.

> This is indeed a big challenge
Yes, it's definitely not an initial goal, but I can see it being useful from an IDE standpoint.

> I think this merits a separate repo, scalameta/scalagen.
Fair, I'll create one soon.

My number 1 priority is writing enough of the code to get extension based macros working. Everything else is irrelevant until then. I also have not fully decided how companion objects will work, it's a rather dangerous area. I think adding items to companion objects and creating companion objects will be fine, but manipulation will be prohibited.

**Comment by olafurpg** [comment]
Behold https://github.com/scalameta/scalagen 

Awaiting your PR @DavidDudson 💃 

> `@data class Point(x: Byte, y: Byte)`

Assuming there is some `object data { def expand(cls: Defn.Class): Defn.Class }`, I think it might be more convenient to call `data.expand(specializedOutput)` from the `@specialized` generator instead of expanding into `@data class ....` and letting scalagen figure out how to expand it. The latter is fragile since you must rely on `@data` resolving to the correct data annotation in that given scope.

> When we are introspecting our enclosing scope, we have to understand what is subject to change.

I may be misunderstanding you, but if we don't look for annotations in the expanded code then we don't need to understand so much what is changed. We can make the expansion methods reusable (plain static methods on objects) so one generator can programmatically invoke other generators without relying on scalagen to compose generators.

> My number 1 priority is writing enough of the code to get extension based macros working. 

Sounds good 👍  I agree it's best to try and get something off the ground that's useful for your application and we can discuss the design better in PRs.

**Comment by DavidDudson** [comment]
> I may be misunderstanding you, but if we don't look for annotations in the expanded code then we don't need to understand so much what is changed. 

It's not for us. It's for users. We have no reason to not let users view the source for a given tree at expansion time. They can happily introspect the entire file, provided they understand the restrictions.

**Comment by DavidDudson** [comment]
I'm going to close this. Any further discussion can be had in scalameta/scalagen
